### ISSUE TITLE
observer for WIFI disconnected state doesn't work

### ISSUE THREAD

**Issue by wujianv5** [pwittchen/ReactiveNetwork#159]
```java
ReactiveNetwork.observeNetworkConnectivity(context)
                .subscribeOn(Schedulers.io())
                .filter(ConnectivityPredicate.hasState(NetworkInfo.State.CONNECTED, NetworkInfo.State.DISCONNECTED))
                .filter(ConnectivityPredicate.hasType(ConnectivityManager.TYPE_WIFI))
                .observeOn(Schedulers.io())
                .subscribe((connectivity) -> {
                    switch (connectivity.getState()) {
                        case CONNECTED:
                            onWifiConnected();
                            break;
                        case DISCONNECTED:
                            onWifiDisconnected();
                            break;
                    }
                });
```

With above code, when I switched off wifi, onWifiDisconnected() wasn't called.
And then I tried to remove .filter(ConnectivityPredicate.hasType(ConnectivityManager.TYPE_WIFI)), and onWifiDisconnected() was called with connectivity.getTypeName() == null, it's supposed to be WIFI.

**Comment by pwittchen** [comment]
Tkanks for reporting that issue. I'll investigate that. I suppose we cannot
filter the network type if we want to detect the state when device is
disconnected, because in that case type probably will be always null. I
mean, when we're offline there's no network type to read.

06.03.2017 6:18 AM "wujian" <notifications@github.com> napisał(a):

ReactiveNetwork.observeNetworkConnectivity(context)
.subscribeOn(Schedulers.io())
.filter(ConnectivityPredicate.hasState(NetworkInfo.State.CONNECTED,
NetworkInfo.State.DISCONNECTED))
.filter(ConnectivityPredicate.hasType(ConnectivityManager.TYPE_WIFI))
.observeOn(Schedulers.io())
.subscribe((connectivity) -> {
switch (connectivity.getState()) {
case CONNECTED:
onWifiConnected();
break;
case DISCONNECTED:
onWifiDisconnected();
break;
}
});

With above code, when I switched off wifi, onWifiDisconnected() wasn't
called.
And then I tried to remove .filter(ConnectivityPredicate.
hasType(ConnectivityManager.TYPE_WIFI)), and onWifiDisconnected() was
called with connectivity.getType() == null, it's supposed to be WIFI.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/pwittchen/ReactiveNetwork/issues/159>, or mute the
thread

**Comment by pwittchen** [comment]
I took a quick look at the code. We can try to keep information about the last network device was connected to and then pass it to `Connectivity` object. This will allow obtaining the desired result.

**Comment by wuasiam** [comment]
I'm not familiar with network state changing things, but I think when I switch off wifi, I'm not in an offline mode, since I still have mobile network connection, and maybe I could also have bluetooth connection. In that cases I guess the disconnection type can be detected ??

BTW, it's a really good work and a very easy tool to use. Thanks!

**Comment by pwittchen** [comment]
You're right. Android SDK works in such way and behavior with filtering Observable mentioned by you may be confusing. It can be fixed within the library code. I have some sort of idea how to do that and I will notify you about the updates in this issue.

**Comment by pwittchen** [comment]
@wujianv5 Can you provide details of your device and Android version? I've performed a test on Nexus 6 with Android 7.0 and it works fine. I mean, bug reported by you does not occur. When I apply filter: `ConnectivityPredicate.hasType(ConnectivityManager.TYPE_WIFI)`, `connectivity.getTypeName()` is never `null`. It does not matter if it's in `CONNECTED` or `DISCONNECTED` state. It returns `WIFI` all the time.

**Comment by wujianv5** [comment]
My phone is MI 5s, and Android version is 6.0.1 MXB48T. I think because it's a customized version of Android, the manufacturer might have changed something.

**Comment by IonutNegru87** [comment]
I encounter the same issue and I think it is because of the `Connectivity.create(context)` in the `NetworkCallback#onLost(Network)`. 
From what I noticed, this will create an Connectivity object which has the type == -1 (ConnectivityManager.NONE)

Changing `.filter(ConnectivityPredicate.hasType(ConnectivityManager.TYPE_WIFI))` to `.filter(ConnectivityPredicate.hasType(ConnectivityManager.TYPE_WIFI, -1))`, seems to fix this issue.

**Comment by pwittchen** [comment]
If the solution provided by @IonutNegru87 works, we can add this type (unknown), to the list of initial types by default inside `ConnectivityPredicate.hasType(...)` method. This should fix problems for all types of connections in the described situation because when we are disconnected from the network, the type is unknown. It's correct behavior but may be misleading in the case of reactive programming when we don't operate on the state, but on the stream.

**Comment by pwittchen** [comment]
I've created an initial fix in PR #186. Now, it's tested only via unit tests. I'll perform a few manual tests before a merge. You can also test it on a feature branch on your own if you would like to. Any feedback is welcome. This PR is for `RxJava2.x` branch, but we can add it to `RxJava1.x` branch as well (e.g. via git cherry-pick).

**Comment by pwittchen** [comment]
Fix will be available in the next release.
